### PR TITLE
MM-29853: Expose plugin-apps REST APIs

### PR DESCRIPTION
**Pull request by catalintomai**
#### Summary
Expose plugin-apps REST APIs

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-29583

### CODE REVIEW

File: modelapps/mattermost_client_pp.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+func (c *ClientPP) DoApiDelete(url string) (*http.Response, *model.AppError) {
+	return c.DoApiRequest(http.MethodDelete, c.ApiUrl+url, "", "")
+}
+
+func (c *ClientPP) DoApiRequest(method, url, data, etag string) (*http.Response, *model.AppError) {
+	return c.doApiRequestReader(method, url, strings.NewReader(data), etag)
+}
+
+func (c *ClientPP) doApiRequestReader(method, url string, data io.Reader, etag string) (*http.Response, *model.AppError) {
```
Comment (catalintomai) [inline modelapps/mattermost_client_pp.go:81]: Would like to re-use the mm-server client4.go methods instead, however, they'll need to be factored out of client4.go first.

File: modelapps/mattermost_client_pp.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+	if len(etag) > 0 {
+		rq.Header.Set(HEADER_ETAG_CLIENT, etag)
+	}
+
+	if len(c.AuthToken) > 0 {
+		rq.Header.Set(HEADER_AUTH, c.AuthType+" "+c.AuthToken)
+	}
+
+	if c.HttpHeader != nil && len(c.HttpHeader) > 0 {
```
Comment (larkox) [inline modelapps/mattermost_client_pp.go:95]: If I am not mistaken, len should return 0 on nil maps, so the first check should not be needed.

File: modelapps/mattermost_client_pp.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	HttpHeader map[string]string // Headers to be copied over for each request
+
+	// TrueString is the string value sent to the server for true boolean query parameters.
+	trueString string
+
+	// FalseString is the string value sent to the server for false boolean query parameters.
+	falseString string
+}
+
+func NewAPIClientPP(url string) *ClientPP {
```
Comment (larkox) [inline modelapps/mattermost_client_pp.go:37]: What does Client PP stands for?

**Review by larkox — APPROVED** [review]
LGTM, just one nit and one question.

File: modelapps/mattermost_client_pp.go (hunk 3 above)
Comment (catalintomai) [reply modelapps/mattermost_client_pp.go:37]: PP-ProxyPlugin (I think Shota introduced it first) - it is the golang client driver for the PP REST API.

**Review by levb — CHANGES_REQUESTED** [review]
1/5 s/modelapps/apps/ and 0/5 move all Client stuff into a separate package (/apps/mmclient?)

this way, `apps.Form` or `apps.Call` is intention-revealing and compact; ditto mmclient.Client. The `api` folder can be adjusted since there's little left in there; separate PR.

File: server/api/impl/appservices/subscriptions.go [hunk 4]
```diff
@@ -3,12 +3,22 @@
...
 
-import "github.com/mattermost/mattermost-plugin-apps/server/api"
+import (
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+)
 
-func (a *AppServices) Subscribe(sub *api.Subscription) error {
-	return a.store.StoreSub(sub)
+func (a *AppServices) Subscribe(sub *apps.Subscription) error {
+	err := a.store.StoreSub(sub)
```
Comment (levb) [inline server/api/impl/appservices/subscriptions.go:11]: nit: why?

**Review by levb — APPROVED** [review]
nice! thanks for some of the cleanup!

File: server/api/impl/appservices/subscriptions.go (hunk 4 above)
Comment (catalintomai) [reply server/api/impl/appservices/subscriptions.go:11]: leftover from debugging, will remove.

**Pull request merged**